cmd/rosetta-dispatcher-server/main.go

```diff
@@ -0,0 +1,297 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	url2 "net/url"
+	"os"
+	"os/signal"
+	"sort"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	"github.com/labstack/echo/v4/middleware"
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/rs/zerolog"
+	"github.com/spf13/pflag"
+	"github.com/ziflex/lecho/v2"
+)
+
+const (
+	success = 0
+	failure = 1
+)
+
+type Spork struct {
+	First      uint64
+	Last       uint64
+	ProxyTaget *middleware.ProxyTarget
+}
+
+type SporkList []Spork
+
+func NewSporkList(addresses []string, firsts []int64, lasts []int64) (SporkList, error) {
+	// TODO: Add tests for validating sporks configuration
+
+	// Check past sporks configuration
+	if len(addresses) != len(firsts) || len(addresses) != len(lasts) {
+		return nil, fmt.Errorf("data length mismatch")
+	}
+
+	if len(addresses) < 2 {
+		return nil, fmt.Errorf("at least two sporks must be provided")
+	}
+
+	sporksTable := make(SporkList, len(addresses))
+
+	for i, sporkAddress := range addresses {
+		sporksTable[i].First = uint64(firsts[i])
+		sporksTable[i].Last = uint64(lasts[i])
+
+		url, err := url2.Parse(sporkAddress)
+		if err != nil {
+			return nil, fmt.Errorf("spork address %d is invalid: %w", i, err)
+		}
+		if sporksTable[i].First >= sporksTable[i].Last {
+			return nil, fmt.Errorf("spork %d last height is not greater than first", i)
+		}
+
+		sporksTable[i].ProxyTaget = &middleware.ProxyTarget{URL: url}
+	}
+
+	sort.Sort(sporksTable)
+
+	for i := 0; i < len(sporksTable)-1; i++ {
+		if sporksTable[i].Last != sporksTable[i+1].First-1 {
+			return nil, fmt.Errorf("gap between sporks boundaries: %d and %d", sporksTable[i].Last, sporksTable[i+1].First)
+		}
+	}
+
+	return sporksTable, nil
+}
+
+func (s SporkList) Len() int {
+	return len(s)
+}
+
+func (s SporkList) Less(i, j int) bool {
+	return s[i].First < s[j].First
+}
+
+func (s SporkList) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func (s SporkList) ServerForHeight(height uint64) (*middleware.ProxyTarget, error) {
+	// Assume sorted and more than zero items
+
+	first := s[0].First
+	last := s[len(s)-1].Last
+
+	if height < first {
+		return nil, fmt.Errorf("height %d below supported range %d - %d", height, first, last)
+	}
+
+	if height > last {
+		return nil, fmt.Errorf("height %d above supported range %d - %d", height, first, last)
+	}
+
+	for _, spork := range s {
+		if height >= spork.First && height <= spork.Last {
+			return spork.ProxyTaget, nil
+		}
+	}
+
+	return nil, fmt.Errorf("spork for height %d not found", height)
+}
+
+func NewFlowHeightAwareBalancer(sporkList SporkList) *FlowHeightAwareBalancer {
+
+	//targets := make([]*middleware.ProxyTarget, sporkList.Len())
+	//
+	//for i, spork := range sporkList {
+	//
+	//}
+
+	return &FlowHeightAwareBalancer{
+		//targets:   nil,
+		sporkList: sporkList,
```

Comment:
From the POV of the middleware, I feel like this should be called `targets`, since the middleware does not care about the details behind what the targets are, IMO. WDYT?